gammapy/cube/fit.py

```diff
@@ -208,22 +209,37 @@ def apply_psf(self, npred):
         """Convolve npred cube with PSF"""
         return npred.convolve(self.psf)
 
-    def apply_edisp(self, data):
-        """Convolve map data with energy dispersion."""
-        data = np.rollaxis(data, 0, 3)
+    def apply_edisp(self, npred):
+        """Convolve map data with energy dispersion.
+
+        npred: npred map in e_true
+        Returns
+        -------
+        npred_reco: npred map in e_reco
+        """
+
+        loc = npred.geom.get_axes_dimension_by_name("energy")
+        data = np.moveaxis(npred.data, loc, -1)
         data = np.dot(data, self.edisp.pdf_matrix)
-        return np.rollaxis(data, 2, 0)
+        data1 = np.moveaxis(data, -1, loc) #now dim is in ereco
```

Comment:
You are modifying `data` and still keep calling the variable `data` on the lines above.
Suggest to continue here and not start calling it `data1`.